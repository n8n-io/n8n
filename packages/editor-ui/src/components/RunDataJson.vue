<template>
	<div :class="$style.jsonDisplay">
		<run-data-json-actions
			v-if="!editMode.enabled"
			:node="node"
			:sessioId="sessionId"
			:displayMode="displayMode"
			:distanceFromActive="distanceFromActive"
			:selectedJsonPath="selectedJsonPath"
			:jsonData="jsonData"
			:paneType="paneType"
		/>
		<draggable
			type="mapping"
			targetDataKey="mappable"
			:disabled="!mappingEnabled"
			@dragstart="onDragStart"
			@dragend="onDragEnd"
			ref="draggable"
		>
			<template #preview="{ canDrop, el }">
				<div :class="[$style.dragPill, canDrop ? $style.droppablePill : $style.defaultPill]">
					{{ $locale.baseText('dataMapping.mapKeyToField', { interpolate: { name: getShortKey(el) } }) }}
				</div>
			</template>
			<template>
				<vue-json-pretty
					:data="jsonData"
					:deep="10"
					:showLength="true"
					:selected-value.sync="selectedJsonPath"
					rootPath=""
					selectableType="single"
					class="json-data"
				>
					<template #nodeKey="{ node }">
					<span
						data-target="mappable"
						:data-value="getJsonParameterPath(node.path)"
						:data-name="node.key"
						:data-path="node.path"
						:data-depth="node.level"
						:class="{
							[$style.mappable]: mappingEnabled,
							[$style.dragged]: draggingPath === node.path,
						}"
					>"{{ node.key }}"</span>
					</template>
					<template #nodeValue="{ node }">
						<span>{{ getContent(node.content) }}</span>
					</template>
				</vue-json-pretty>
			</template>
		</draggable>
	</div>
</template>

<script lang="ts">
import { PropType } from "vue";
import mixins from "vue-typed-mixins";
import VueJsonPretty from 'vue-json-pretty';
import { LOCAL_STORAGE_MAPPING_FLAG } from '@/constants';
import { IDataObject, INodeExecutionData } from "n8n-workflow";
import Draggable from '@/components/Draggable.vue';
import { convertPath, executionDataToJson, isString, isStringNumber } from "@/components/helpers";
import { INodeUi } from "@/Interface";
import { shorten } from './helpers';
import { externalHooks } from "@/components/mixins/externalHooks";

const runDataJsonActions = () => import('@/components/RunDataJsonActions.vue');

export default mixins(externalHooks).extend({
	name: 'run-data-json',
	components: {
		VueJsonPretty,
		Draggable,
		runDataJsonActions,
	},
	props: {
		editMode: {
			type: Object as () => { enabled?: boolean; value?: string; },
		},
		currentOutputIndex: {
			type: Number,
		},
		sessionId: {
			type: String,
		},
		paneType: {
			type: String,
		},
		node: {
			type: Object as PropType<INodeUi>,
		},
		inputData: {
			type: Array as PropType<INodeExecutionData[]>,
		},
		mappingEnabled: {
			type: Boolean,
		},
		distanceFromActive: {
			type: Number,
		},
		showMappingHint: {
			type: Boolean,
		},
		runIndex: {
			type: Number,
		},
		totalRuns: {
			type: Number,
		},
	},
	data() {
		return {
			selectedJsonPath: null as null | string,
			mappingHintVisible: false,
			showHintWithDelay: false,
			draggingPath: null as null | string,
			displayMode: 'json',
		};
	},
	mounted() {
		if (this.showMappingHint) {
			this.mappingHintVisible = true;

			setTimeout(() => {
				this.mappingHintVisible = false;
			}, 6000);
		}

		if (this.showMappingHint && this.showHint) {
			setTimeout(() => {
				this.showHintWithDelay = this.showHint;
				this.$telemetry.track('User viewed JSON mapping tooltip', { type: 'param focus' });
			}, 500);
		}
	},
	computed: {
		jsonData(): IDataObject[] {
			return executionDataToJson(this.inputData as INodeExecutionData[]);
		},
		showHint(): boolean {
			return (
				!this.draggingPath &&
				((this.showMappingHint && this.mappingHintVisible) ||
					window.localStorage.getItem(LOCAL_STORAGE_MAPPING_FLAG) !== 'true')
			);
		},
	},
	methods: {
		getShortKey(el: HTMLElement): string {
			if (!el) {
				return '';
			}

			return shorten(el.dataset.name || '', 16, 2);
		},
		getJsonParameterPath(path: string): string {
			const convertedPath = convertPath(path);
			return `{{ ${ convertedPath.replace(/^(\["?\d"?])/, this.distanceFromActive === 1 ? '$json' : `$node["${ this.node!.name }"].json`) } }}`;
		},
		onDragStart(el: HTMLElement) {
			if (el && el.dataset.path) {
				this.draggingPath = el.dataset.path;
			}

			this.$store.commit('ui/resetMappingTelemetry');
		},
		onDragEnd(el: HTMLElement) {
			this.draggingPath = null;

			setTimeout(() => {
				const mappingTelemetry = this.$store.getters['ui/mappingTelemetry'];
				const telemetryPayload = {
					src_node_type: this.node.type,
					src_field_name: el.dataset.name || '',
					src_nodes_back: this.distanceFromActive,
					src_run_index: this.runIndex,
					src_runs_total: this.totalRuns,
					src_field_nest_level: el.dataset.depth || 0,
					src_view: 'json',
					src_element: el,
					success: false,
					...mappingTelemetry,
				};

				this.$externalHooks().run('runDataJson.onDragEnd', telemetryPayload);

				this.$telemetry.track('User dragged data for mapping', telemetryPayload);
			}, 1000); // ensure dest data gets set if drop
		},
		getContent(value: string): string {
			return isString(value) && !isStringNumber(value) ? `"${ value }"` : value;
		},
	},
});
</script>

<style lang="scss" module>
.jsonDisplay {
	position: absolute;
	top: 0;
	left: 0;
	padding-left: var(--spacing-s);
	right: 0;
	overflow-y: auto;
	line-height: 1.5;
	word-break: normal;
	height: 100%;
	padding-bottom: var(--spacing-3xl);
	background-color: var(--color-background-base);
	padding-top: var(--spacing-s);

	&:hover {
		/* Shows .actionsGroup element from <run-data-json-actions /> child component */
		> div:first-child {
			opacity: 1;
		}
	}
}

.mappable {
	cursor: grab;

	&:hover {
		background-color: var(--color-json-highlight);
	}
}

.dragged {
	&,
	&:hover {
		background-color: var(--color-primary-tint-2);
	}
}

.dragPill {
	padding: var(--spacing-4xs) var(--spacing-4xs) var(--spacing-3xs) var(--spacing-4xs);
	color: var(--color-text-xlight);
	font-weight: var(--font-weight-bold);
	font-size: var(--font-size-2xs);
	border-radius: var(--border-radius-base);
	white-space: nowrap;
}

.droppablePill {
	background-color: var(--color-success);
}

.defaultPill {
	background-color: var(--color-primary);
	transform: translate(-50%, -100%);
	box-shadow: 0 2px 6px rgba(68, 28, 23, 0.2);
}

</style>

<style lang="scss">
.vjs-tree {
	color: var(--color-json-default);
}

.vjs-tree-node {
	&:hover {
		background-color: transparent;
	}

	&.is-highlight {
		background-color: var(--color-json-highlight);
	}
}

.vjs-key,
.vjs-value {
	> span {
		color: var(--color-text-dark);
		line-height: 1.7;
		border-radius: var(--border-radius-base);
	}
}

.vjs-value {
	> span {
		padding: 0 var(--spacing-5xs) 0 var(--spacing-5xs);
		margin-left: var(--spacing-5xs);
	}
}


.vjs-tree .vjs-value-null {
	&, span {
		color: var(--color-json-null);
	}
}

.vjs-tree .vjs-value-boolean {
	&, span {
		color: var(--color-json-boolean);
	}
}

.vjs-tree .vjs-value-number {
	&, span {
		color: var(--color-json-number);
	}
}

.vjs-tree .vjs-value-string {
	&, span {
		color: var(--color-json-string);
	}
}

.vjs-tree .vjs-key {
	color: var(--color-json-key);
}

.vjs-tree .vjs-tree__brackets {
	color: var(--color-json-brackets);
}

.vjs-tree .vjs-tree__brackets:hover {
	color: var(--color-json-brackets-hover);
}

.vjs-tree .vjs-tree__content.has-line {
	border-left: 1px dotted var(--color-json-line);
}
</style>
