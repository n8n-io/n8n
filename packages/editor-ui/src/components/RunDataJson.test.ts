import Vue from 'vue';
import { PiniaVuePlugin } from 'pinia';
import { createTestingPinia } from '@pinia/testing';
import { render, screen, cleanup } from '@testing-library/vue';
import RunDataJson from '@/components/RunDataJson.vue';

Vue.use(PiniaVuePlugin);

describe('RunDataJson.vue', () => {
	const DEFAULT_SETUP = {
		pinia: createTestingPinia(),
		props: {
			mappingEnabled: true,
			editMode: { enabled: false },
			inputData: [
				{
					json: {
						list: [1, 2, 3],
						record: { name: 'Joe' },
						myNumber: 123,
						myStringNumber: '456',
						myStringText: 'abc',
						nil: null,
						d: undefined,
					},
				},
			],
			node: {
				parameters: {
					keepOnlySet: false,
					values: {},
					options: {},
				},
				id: '820ea733-d8a6-4379-8e73-88a2347ea003',
				name: 'Set',
				type: 'n8n-nodes-base.set',
				typeVersion: 1,
				position: [380, 1060],
				disabled: false,
			},
		},
		mocks: {
			$locale: {
				baseText() {
					return '';
				},
			},
			$store: {
				getters: {},
			},
		},
	};

	beforeEach(cleanup);

	it('renders json values properly', () => {
		const { container } = render(RunDataJson, DEFAULT_SETUP, (vue) => {
			vue.use(PiniaVuePlugin);
		});
		expect(container).toMatchSnapshot();

		expect(screen.getByText('123')).toBeInTheDocument();
		expect(screen.getByText('"456"')).toBeInTheDocument();
		expect(screen.getByText('"abc"')).toBeInTheDocument();
		expect(screen.getByText('null')).toBeInTheDocument();
		expect(screen.queryByText('undefined')).not.toBeInTheDocument();
	});

	it('sets ph-no-capture class correctly', () => {
		render(RunDataJson, DEFAULT_SETUP);

		expect(screen.getByText('"list"')).not.toHaveClass('ph-no-capture');
		expect(screen.getByText('"record"')).not.toHaveClass('ph-no-capture');
		expect(screen.getByText('"myStringNumber"')).not.toHaveClass('ph-no-capture');

		expect(screen.getByText('123')).toHaveClass('ph-no-capture');
		expect(screen.getByText('"456"')).toHaveClass('ph-no-capture');
		expect(screen.getByText('"abc"')).toHaveClass('ph-no-capture');
		expect(screen.getByText('null')).toHaveClass('ph-no-capture');
	});
});
