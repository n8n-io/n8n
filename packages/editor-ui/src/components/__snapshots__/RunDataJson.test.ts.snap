// Vitest Snapshot v1

exports[`RunDataJson.vue > renders json values properly 1`] = `
<div>
  <div
    class="jsonDisplay"
  >
    <!---->
    <div
      class=""
    >
      <div
        class="json-data vjs-tree"
      >
        <div
          class="vjs-tree-list"
        >
          <div
            class="vjs-tree-list-holder"
          >
            <div
              class="vjs-tree-list-holder-inner"
            >
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-tree-brackets"
                  >
                    [
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-tree-brackets"
                  >
                    {
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <span
                  class="vjs-key"
                >
                  <span
                    class="mappable"
                    data-depth="2"
                    data-name="list"
                    data-path="[0].list"
                    data-target="mappable"
                    data-value="{{ $node.Set.json.list }}"
                  >
                    "list"
                  </span>
                  <span>
                    :
                  </span>
                </span>
                <span>
                  <span
                    class="vjs-tree-brackets"
                  >
                    [
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-value vjs-value-number"
                  >
                    <span
                      class="ph-no-capture mappable"
                      data-depth="3"
                      data-name="list[0]"
                      data-path="[0].list[0]"
                      data-target="mappable"
                      data-value="{{ $node.Set.json.list[0] }}"
                    >
                      1
                    </span>
                  </span>
                  <span>
                    ,
                  </span>
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-value vjs-value-number"
                  >
                    <span
                      class="ph-no-capture mappable"
                      data-depth="3"
                      data-name="list[1]"
                      data-path="[0].list[1]"
                      data-target="mappable"
                      data-value="{{ $node.Set.json.list[1] }}"
                    >
                      2
                    </span>
                  </span>
                  <span>
                    ,
                  </span>
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-value vjs-value-number"
                  >
                    <span
                      class="ph-no-capture mappable"
                      data-depth="3"
                      data-name="list[2]"
                      data-path="[0].list[2]"
                      data-target="mappable"
                      data-value="{{ $node.Set.json.list[2] }}"
                    >
                      3
                    </span>
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-tree-brackets"
                  >
                    ]
                  </span>
                  <span>
                    ,
                  </span>
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <span
                  class="vjs-key"
                >
                  <span
                    class="mappable"
                    data-depth="2"
                    data-name="record"
                    data-path="[0].record"
                    data-target="mappable"
                    data-value="{{ $node.Set.json.record }}"
                  >
                    "record"
                  </span>
                  <span>
                    :
                  </span>
                </span>
                <span>
                  <span
                    class="vjs-tree-brackets"
                  >
                    {
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <span
                  class="vjs-key"
                >
                  <span
                    class="mappable"
                    data-depth="3"
                    data-name="name"
                    data-path="[0].record.name"
                    data-target="mappable"
                    data-value="{{ $node.Set.json.record.name }}"
                  >
                    "name"
                  </span>
                  <span>
                    :
                  </span>
                </span>
                <span>
                  <span
                    class="vjs-value vjs-value-string"
                  >
                    <span
                      class="ph-no-capture"
                    >
                      "Joe"
                    </span>
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-tree-brackets"
                  >
                    }
                  </span>
                  <span>
                    ,
                  </span>
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <span
                  class="vjs-key"
                >
                  <span
                    class="mappable"
                    data-depth="2"
                    data-name="myNumber"
                    data-path="[0].myNumber"
                    data-target="mappable"
                    data-value="{{ $node.Set.json.myNumber }}"
                  >
                    "myNumber"
                  </span>
                  <span>
                    :
                  </span>
                </span>
                <span>
                  <span
                    class="vjs-value vjs-value-number"
                  >
                    <span
                      class="ph-no-capture"
                    >
                      123
                    </span>
                  </span>
                  <span>
                    ,
                  </span>
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <span
                  class="vjs-key"
                >
                  <span
                    class="mappable"
                    data-depth="2"
                    data-name="myStringNumber"
                    data-path="[0].myStringNumber"
                    data-target="mappable"
                    data-value="{{ $node.Set.json.myStringNumber }}"
                  >
                    "myStringNumber"
                  </span>
                  <span>
                    :
                  </span>
                </span>
                <span>
                  <span
                    class="vjs-value vjs-value-string"
                  >
                    <span
                      class="ph-no-capture"
                    >
                      "456"
                    </span>
                  </span>
                  <span>
                    ,
                  </span>
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <span
                  class="vjs-key"
                >
                  <span
                    class="mappable"
                    data-depth="2"
                    data-name="myStringText"
                    data-path="[0].myStringText"
                    data-target="mappable"
                    data-value="{{ $node.Set.json.myStringText }}"
                  >
                    "myStringText"
                  </span>
                  <span>
                    :
                  </span>
                </span>
                <span>
                  <span
                    class="vjs-value vjs-value-string"
                  >
                    <span
                      class="ph-no-capture"
                    >
                      "abc"
                    </span>
                  </span>
                  <span>
                    ,
                  </span>
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <span
                  class="vjs-key"
                >
                  <span
                    class="mappable"
                    data-depth="2"
                    data-name="nil"
                    data-path="[0].nil"
                    data-target="mappable"
                    data-value="{{ $node.Set.json.nil }}"
                  >
                    "nil"
                  </span>
                  <span>
                    :
                  </span>
                </span>
                <span>
                  <span
                    class="vjs-value vjs-value-null"
                  >
                    <span
                      class="ph-no-capture"
                    >
                      null
                    </span>
                  </span>
                  <span>
                    ,
                  </span>
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <span
                  class="vjs-key"
                >
                  <span
                    class="mappable"
                    data-depth="2"
                    data-name="d"
                    data-path="[0].d"
                    data-target="mappable"
                    data-value="{{ $node.Set.json.d }}"
                  >
                    "d"
                  </span>
                  <span>
                    :
                  </span>
                </span>
                <span>
                  <span
                    class="vjs-value vjs-value-undefined"
                  >
                    <span
                      class="ph-no-capture"
                    >
                      
                    </span>
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <div
                    class="vjs-indent-unit has-line"
                  />
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-tree-brackets"
                  >
                    }
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
              <div
                class="vjs-tree-node"
                collapsed-on-click-brackets="true"
              >
                <!---->
                <!---->
                <div
                  class="vjs-indent"
                >
                  <!---->
                </div>
                <!---->
                <span>
                  <span
                    class="vjs-tree-brackets"
                  >
                    ]
                  </span>
                  <!---->
                  <!---->
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="teleporter hidden"
        data-v-d4e6e290=""
      />
    </div>
  </div>
</div>
`;
